src/Sylius/Behat/Context/Setup/CatalogPromotionContext.php

```diff
@@ -756,13 +758,34 @@ public function theCatalogPromotionPriorityIs(CatalogPromotionInterface $catalog
         $this->eventBus->dispatch(new CatalogPromotionUpdated($catalogPromotion->getCode()));
     }
 
+    /**
+     * @Given /^the ("[^"]+" catalog promotion) is (active|inactive)$/
+     * @Given /^(this catalog promotion) is (active|inactive)$/
+     */
+    public function theCatalogPromotionIs(CatalogPromotionInterface $catalogPromotion, string $state)
+    {
+        $this->setCatalogPromotionState($catalogPromotion, $state);
+    }
+
     /**
      * @Given the catalog promotion :catalogPromotion is currently being processed
      */
     public function theCatalogPromotionIsCurrentlyBeingProcessed(CatalogPromotionInterface $catalogPromotion): void
     {
-        $stateMachine = $this->stateMachineFactory->get($catalogPromotion, CatalogPromotionTransitions::GRAPH);
-        $stateMachine->apply(CatalogPromotionTransitions::TRANSITION_PROCESS);
+        $this->setCatalogPromotionState($catalogPromotion, CatalogPromotionStates::STATE_PROCESSING);
+    }
+
+    /**
+     * @Given the processing of :catalogPromotion has failed
+     */
+    public function theProcessingOfCatalogPromotionHasFailed(CatalogPromotionInterface $catalogPromotion): void
+    {
+        $this->setCatalogPromotionState($catalogPromotion, CatalogPromotionStates::STATE_FAILED);
+    }
+
+    private function setCatalogPromotionState(CatalogPromotionInterface $catalogPromotion, string $state): void
+    {
+        $catalogPromotion->setState($state);
```

Comment:
Please, no :( If we manipulate catalog promotions' states, we should definitely use state machine transitions to achieve them. I know, we have such abomination already in `OrderContext`, but we should definitely get rid of them 🖖